test/functional/p2p_filter.py

```diff
@@ -51,6 +56,21 @@ def on_merkleblock(self, message):
     def on_tx(self, message):
         self.tx_received = True
 
+class FilterNodeFRelay(FilterNode):
+    def peer_connect(self, *args, services=NODE_NETWORK|NODE_WITNESS, send_version=True, **kwargs):
+        create_conn = P2PConnection.peer_connect(self, *args, **kwargs)
+
+        if send_version:
+            # Send a version msg
+            vt = msg_version()
+            vt.nServices = services
+            vt.addrTo.ip = self.dstaddr
+            vt.addrTo.port = self.dstport
+            vt.addrFrom.ip = "0.0.0.0"
+            vt.addrFrom.port = 0
+            vt.nRelay = 0 # default = 1
+            self.on_connection_send_msg = vt  # Will be sent soon after connection_made
+        return create_conn
```

Comment:
Is there any reason to overwrite this at all? The whole overwrite with the `on_connection_send_msg` magic is a bit of a mental load. I believe this can simply be a `FilterNode`, where you set `send_version=False` and then `send_message(version_without_nRelay)` and then `wait_for_verack()`.

Overall it should be less complex and less lines of code

Reply:
Oops - had a big hole in my knowledge. How does this look:
```py
class FilterNodeWithoutNRelay(FilterNode):
    def __init__(self):
        super().__init__()
        version_without_nRelay = msg_version()
        version_without_nRelay.nRelay = 0
        self.on_connection_send_msg = version_without_nRelay
```
I'm still not 100% clear on how connection works under the hood, but setting `self.on_connection_send_msg` feels correct since it's handled in the `connection_made` callback - otherwise I need to somehow wait for connection to be established before using `send_message` directly.

Reply:
When I added `on_connection_send_msg` it was meant to be a "private" member. Obviously in python there is no "private", so discount my point by that.

I think either is fine. Your solution with the `on_connection_send_msg` or something along the lines of:

```
p2p = create_connection(bla)
p2p.wait_until(p2p.is_connected)
p2p.send_message(ver)
p2p.wait_for_verack()